tests/test_scarf_telemetry.py

```diff
@@ -0,0 +1,147 @@
+    os.environ["SCARF_NO_ANALYTICS"] = "true"
+
+    response_status, runner_type = scarf_telemetry(runner="ray")
+
+    assert response_status is None
+    assert runner_type is None
+    mock_urlopen.assert_not_called()
+
+
+@patch("daft.scarf_telemetry.get_build_type")
+@patch("daft.scarf_telemetry.get_version")
+@patch("urllib.request.urlopen")
+def test_scarf_telemetry_opt_out_with_do_not_track(
+    mock_urlopen: MagicMock,
+    mock_version: MagicMock,
+    mock_build_type: MagicMock,
+):
+    mock_version.return_value = "0.0.0"
+    mock_build_type.return_value = "release"
+
+    # Clean and set environment
+    os.environ.pop("SCARF_NO_ANALYTICS", None)
+    os.environ["DO_NOT_TRACK"] = "true"
+
+    response_status, runner_type = scarf_telemetry(runner="ray")
+
+    assert response_status is None
+    assert runner_type is None
+    mock_urlopen.assert_not_called()
+
+
+@patch("daft.scarf_telemetry.get_build_type")
+@patch("daft.scarf_telemetry.get_version")
+@patch("urllib.request.urlopen")
+def test_scarf_telemetry_error_handling(
+    mock_urlopen: MagicMock,
+    mock_version: MagicMock,
+    mock_build_type: MagicMock,
+):
+    # Test error handling in scarf_telemetry, verifies that network errors are caught, function returns error message and None for runner type
+
+    # Clean environment
+    os.environ.pop("SCARF_NO_ANALYTICS", None)
+    os.environ.pop("DO_NOT_TRACK", None)
+
+    mock_version.return_value = "0.0.0"
+    mock_build_type.return_value = "release"
+    mock_urlopen.side_effect = urllib.error.URLError(socket.timeout("Timeout"))
+
+    response_status, runner_type = scarf_telemetry(runner="ray")
+
+    assert response_status.startswith("Analytics error:")
+    assert runner_type is None
+
+
+# Tests for runner integration with scarf_telemetry, commented out because cannot set runner more than once
+# @patch("daft.context.scarf_telemetry")
+# def test_runner_ray_analytics(mock_scarf_telemetry: MagicMock):
+#     set_runner_ray()
+#     mock_scarf_telemetry.assert_called_once_with(runner="ray")
+
+# @patch("daft.context.scarf_telemetry")
+# def test_runner_py_analytics(mock_scarf_telemetry: MagicMock):
+#     set_runner_py()
+#     mock_scarf_telemetry.assert_called_once_with(runner="py")
+
+# @patch("daft.context.scarf_telemetry")
+# def test_runner_native_analytics(mock_scarf_telemetry: MagicMock):
+#     set_runner_native()
+#     mock_scarf_telemetry.assert_called_once_with(runner="native")
+
+# @patch("daft.context.scarf_telemetry")
+# def test_runner_analytics_with_scarf_opt_out(mock_scarf_telemetry: MagicMock):
+#     os.environ["SCARF_NO_ANALYTICS"] = "true"
+#     set_runner_ray()
+#     mock_scarf_telemetry.assert_called_once_with(runner="ray")
+
+# @patch("daft.context.scarf_telemetry")
+# def test_runner_analytics_with_do_not_track(mock_scarf_telemetry: MagicMock):
+#     os.environ["DO_NOT_TRACK"] = "true"
+#     set_runner_ray()
+#     mock_scarf_telemetry.assert_called_once_with(runner="ray")
```

Comment:
@universalmind303,

@kevinzwang mentioned yesterday that some of his changes to `daft-context` enabled the ability to set the runner multiple times. But given our move to implementing context on the Rust side instead, is that behaviour still available?

Reply:
Oh for testing, I do have a way to do it, but it's jank. Here's the code that I have: https://github.com/Eventual-Inc/Daft/blob/main/tests/actor_pool/test_actor_cuda_devices.py#L21-L43